salt/modules/x509.py

```diff
@@ -1392,18 +1393,43 @@ def create_certificate(
                 ['listen_in', 'preqrequired', '__prerequired__']:
             kwargs.pop(ignore, None)
 
-        certs = __salt__['publish.publish'](
-            tgt=ca_server,
-            fun='x509.sign_remote_certificate',
-            arg=six.text_type(kwargs))
+        # If the provided argument is not a list, we assume it only contain one server
+        if not isinstance(ca_server, list):
+            ca_server = [ca_server]
+        random.shuffle(ca_server)
+
+        # This variable will contains some errors logs collected during the loop
+        # over the CA servers provided, it will be return if no certificate got generated
+        error_log = []
+        for server in ca_server:
+            certs = __salt__['publish.publish'](
+                tgt=server,
+                fun='x509.sign_remote_certificate',
+                arg=six.text_type(kwargs))
+            if certs is None or not any(certs):
+                error_log.append("{0}: ca_server did not respond "
+                                 "or salt master must permit peers to "
+                                 "call the sign_remote_certificate function.".format(server))
+            else:
+                # Sometimes, an error is returned instead of the certificate.
+
+                # If we provide multiple CA servers, we want to make it HA, so
+                # we have to detect that's not a certificate and try an other server.
+
+                # Also, we don't want to break in case of the cert_txt is a dict
+                # as it could be the output of the "dry-run" first request.
+                cert_txt = certs[server]
+                if isinstance(cert_txt, six.string_types):
+                    if cert_txt.startswith('-----'):
```

Comment:
Could we either use `pyopenssl` or otherwise actually try to check that this *is* a certificate (e.g. use a tempdir and run `openssl x509 ...` on it)? Obviously it's unlikely that anything else is returning `-----`, but I'd feel a lot better if we were making it a *proper* check.

Reply:
Indeed, it would be much cleaner. Other code area are doing this type of test, for example, the "get_pem_entry" function which make this check using a similar (but with a more developed approach). I think, the best would be to a global refactor this module and one of the tasks would be to create a separate and reusable function to check the validity of a certificate. I don't think I would have the time to do a refactor (for now). But this would clearly be a good improvement.